pkg/controller/persistentpodstate/persistent_pod_state_controller.go

```diff
@@ -0,0 +1,451 @@
+		return nil
+	}
+	// fetch persistentPodState crd
+	oldObj := &appsv1alpha1.PersistentPodState{}
+	err = r.Client.Get(context.TODO(), client.ObjectKey{Namespace: ns, Name: name}, oldObj)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return err
+		}
+		// not found
+		oldObj = nil
+	}
+
+	// auto generate persistentPodState crd object
+	if workload.Metadata.Annotations[appsv1alpha1.AnnotationAutoGeneratePersistentPodState] == "true" {
+		if workload.Metadata.Annotations[appsv1alpha1.AnnotationRequiredPersistentTopology] == "" &&
+			workload.Metadata.Annotations[appsv1alpha1.AnnotationPreferredPersistentTopology] == "" {
+			klog.Warningf("statefulSet(%s/%s) persistentPodState annotation is incomplete", workload.Metadata.Namespace, workload.Name)
+			return nil
+		}
+
+		newObj := newStatefulSetPersistentPodState(workload)
+		// create new obj
+		if oldObj == nil {
+			if err := r.Create(context.TODO(), newObj); err != nil && !errors.IsAlreadyExists(err) {
+				return err
+			}
+			klog.V(3).Infof("create StatefulSet(%s/%s) persistentPodState(%s) success", ns, name, util.DumpJSON(newObj))
+			return nil
+		}
+		// compare with old object
+		if reflect.DeepEqual(oldObj.Spec, newObj.Spec) {
+			return nil
+		}
+		objClone := &appsv1alpha1.PersistentPodState{}
+		if err = retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+			if err = r.Client.Get(context.TODO(), client.ObjectKey{Namespace: newObj.Namespace, Name: newObj.Name}, objClone); err != nil {
+				return err
+			}
+			objClone.Spec = *newObj.Spec.DeepCopy()
+			return r.Client.Update(context.TODO(), objClone)
+		}); err != nil {
+			klog.Errorf("update persistentPodState(%s/%s) failed: %s", newObj.Namespace, newObj.Name, err.Error())
+			return err
+		}
+		klog.V(3).Infof("update persistentPodState(%s/%s) from(%s) -> to(%s) success", newObj.Namespace, newObj.Name,
+			util.DumpJSON(oldObj.Spec), util.DumpJSON(newObj.Spec))
+		return nil
+	}
+
+	// delete auto generated persistentPodState crd object
+	if oldObj == nil {
+		return nil
+	}
+	if err = r.Delete(context.TODO(), oldObj); err != nil {
+		return err
+	}
+	klog.V(3).Infof("delete StatefulSet(%s/%s) persistentPodState done", ns, name)
+	return nil
+}
+
+func newStatefulSetPersistentPodState(workload *controllerfinder.ScaleAndSelector) *appsv1alpha1.PersistentPodState {
+	obj := &appsv1alpha1.PersistentPodState{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      workload.Name,
+			Namespace: workload.Metadata.Namespace,
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion: workload.APIVersion,
+					Kind:       workload.Kind,
+					Name:       workload.Name,
+					Controller: utilpointer.BoolPtr(true),
+					UID:        workload.UID,
+				},
+			},
+		},
+		Spec: appsv1alpha1.PersistentPodStateSpec{
+			TargetReference: appsv1alpha1.TargetReference{
+				APIVersion: workload.APIVersion,
+				Kind:       workload.Kind,
+				Name:       workload.Name,
```

Comment:
is it possible that a statefulset and a kruise statefulset with the same name may have conflicting generated pps?